node/pkg/watchers/solana/shim.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", topLevelIdx, err)
+	}
+
+	return true, nil
+}
+
+// shimProcessInnerInstruction handles an inner instruction where the program ID matches the shim contract. It does the following:
+// - Verifies that the instruction is a shim PostMessage. If not, it just returns. If it is, it parses it.
+// - Searches through the subsequent inner instructions in the set to find the entry for the core. Makes sure it is unreliable with no payload.
+// - Searches for the inner shim MessageEvent to get the remaining fields needed to generate the observation.
+// - Publishes the observation.
+func (s *SolanaWatcher) shimProcessInnerInstruction(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) (bool, error) {
+	// See if this is a PostMessage event from the shim contract. If so, parse it. If not, bail out now.
+	postMessage, err := shimParsePostMessage(s.shimPostMessageDiscriminator, innerInstructions[startIdx].Data)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse inner shim post message instruction %d, %d: %w", outerIdx, startIdx, err)
+	}
+
+	if postMessage == nil {
+		return false, nil
+	}
+
+	alreadyProcessed.add(outerIdx, startIdx)
+
+	err = s.shimProcessRest(logger, whProgramIndex, shimProgramIndex, tx, innerInstructions, outerIdx, startIdx+1, postMessage, alreadyProcessed, isReobservation)
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", outerIdx, err)
+	}
+
+	return true, nil
+}
+
+func (s *SolanaWatcher) shimProcessRest(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	postMessage *ShimPostMessageData,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) error {
+	// Loop through the inner instructions after the shim PostMessage and do two things:
+	// 1) Find the core event and verify it is unreliable with an empty payload.
+	// 2) Find the shim MessageEvent to get the rest of the fields we need for the observation.
+
+	var verifiedCoreEvent bool
+	var messageEvent *ShimMessageEventData
+	var err error
+	coreEventFound := false
+	for idx := startIdx; idx < len(innerInstructions); idx++ {
+		inst := innerInstructions[idx]
+		if inst.ProgramIDIndex == whProgramIndex {
+			if verifiedCoreEvent, err = shimVerifyCoreMessage(inst.Data); err != nil {
+				return fmt.Errorf("failed to verify inner core instruction for top-level shim instruction %d: %w", outerIdx, err)
```

Comment:
these errors say "top-level" but I think this function is now used in either case.